### ISSUE TITLE
Unexpected result of "introduce local for all occurrences" refactoring 

### ISSUE THREAD

**Issue by AlekseyTs** [dotnet/roslyn#24521]
Start with this code:
```
public class Class1
{
    Class1 Next;

    public void DoIt(Class1 input)
    {
        Class1 current = input;

        while (true)
        {
            System.Console.WriteLine(current.ToString());

            if (current.Next == null)
            {
                break;
            }

            current = current.Next;
        }

        System.Console.WriteLine(current.ToString());
    }
}
```

Select ```current.ToString()``` within the while loop and do "introduce local for all occurrences" refactoring 

Observed result:
```
        Class1 current = input;
        string value = current.ToString();
        while (true)
        {
            System.Console.WriteLine(value);

            if (current.Next == null)
            {
                break;
            }

            current = current.Next;
        }

        System.Console.WriteLine(value);
```

The result is not semantically equivalent to the original code, but it is possible that I have wrong expectations.

Expected:
```
        Class1 current = input;
        string value;
        while (true)
        {
            value = current.ToString();
            System.Console.WriteLine(value);

            if (current.Next == null)
            {
                break;
            }

            current = current.Next;
        }

        System.Console.WriteLine(value);
```

**Comment by CyrusNajmabadi** [comment]
Yes.  That's the intended way for it to work.  By definition, introducing a local for all occurrences may always change the semantics (unless i guess it was a constant).  The purpose of the feature is to help speed things up when you know you have to make a ton of these changes and you know its safe to do so.  it's not intended to be semantic-changing free.

--

A simpler example is just this:

```c#
Foo(Bar(), Baz());
```

Introducing a local for 'Baz()' is also semantic changing.  But the point is not to be semantic-changing free, but to just speed up the user when they just want to make local and don't want to have to go cut/paste/type-a-bunch/etc.

**Comment by CyrusNajmabadi** [comment]
Note: it would be acceptable to me if the IDE gave a warning icon for these features if every doing something not provably safe.  Though, IMO, the rate of warnings might be so high as to be noisy and not actually provide value in many common cases.

**Comment by AlekseyTs** [comment]
I think that there is something that can be done to improve the "accuracy" of this refactoring, even if we are not shooting for semantic equivalence. For example, references to the expression inside ```if```, loop statements and the like, i.e. constructs that represent control flow, should trigger an assignment to the local next to that reference. For the scenario above that would mean that the code after the refactoring would look like this:
```
        Class1 current = input;
        string value = current.ToString();
        while (true)
        {
            value = current.ToString();
            System.Console.WriteLine(value);

            if (current.Next == null)
            {
                break;
            }

            current = current.Next;
        }

        System.Console.WriteLine(value);
```

**Comment by CyrusNajmabadi** [comment]
I would be fine with the feature offering a *third* option.  something like "introduce local for all in nearest  scope" (or something like that) if there was a difference.   I think there's still value in the current behavior.  

In other words, offer two or three options:

1. Introduce local for expression.
2. Introduce local for all occurrences in scope.
3. Introduce local for all occurrences.

'2' would only be offered if it did something different than '3'.
